selfdrive/controls/tests/test_state_machine.py

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python3
+import unittest
+
+from cereal import car, log
+from common.realtime import DT_CTRL
+from selfdrive.car.car_helpers import interfaces
+from selfdrive.controls.controlsd import Controls, SOFT_DISABLE_TIME
+from selfdrive.controls.lib.events import Events, ET, Alert, Priority, AlertSize, AlertStatus, VisualAlert, \
+                                          AudibleAlert, EVENTS
+
+State = log.ControlsState.OpenpilotState
+
+# The event types that maintain the current state
+MAINTAIN_STATES = {State.enabled: None, State.disabled: None, State.softDisabling: ET.SOFT_DISABLE,
+                   State.preEnabled: ET.PRE_ENABLE}
+ALL_STATES = (state for state in State.schema.enumerants.values() if
+              state != State.overriding)  # TODO: remove overriding exception
+# The event types checked in DISABLED section of state machine
+ENABLE_EVENT_TYPES = (ET.ENABLE, ET.PRE_ENABLE)
+
+
+def make_event(event_types):
+  event = {}
+  for ev in event_types:
+    event[ev] = Alert("", "", AlertStatus.userPrompt, AlertSize.small, Priority.LOW,
+                      VisualAlert.steerRequired, AudibleAlert.prompt, 1.)
+  EVENTS[0] = event
+  return 0
+
+
+class TestStateMachine(unittest.TestCase):
+
+  def setUp(self):
+    CarInterface, CarController, CarState = interfaces["mock"]
+    CP = CarInterface.get_params("mock")
+    CI = CarInterface(CP, CarController, CarState)
+
+    self.controlsd = Controls(CI=CI)
+    self.controlsd.events = Events()
+    self.controlsd.soft_disable_timer = int(SOFT_DISABLE_TIME / DT_CTRL)
+    self.CS = car.CarState()
+
+  def test_immediate_disable(self):
+    for state in ALL_STATES:
+      self.controlsd.events.add(make_event([MAINTAIN_STATES[state], ET.IMMEDIATE_DISABLE]))
```

Comment:
Have to modify the global EVENTS dict to make the Events class use our custom alert. Works, however if you comment any lines like this, then re-run the test, it's passing until you comment every _other_ test for some reason.

Reply:
check out `unittest.mock.patch` https://github.com/commaai/openpilot/blob/b2b6205e252b5291a953dd94c5524715b1f8f444/selfdrive/thermald/tests/test_power_monitoring.py#L3

Reply:
Ah my issue was because wrapping a list comprehension with parenthesis yields a generator, not a tuple as I thought, so it was only iterating through ALL_STATES only in the first test.